File: test/VbNet.LanguageServer.Tests/Integration/CancelRequestIntegrationTests.cs
using System.Text.Json;
using System.Threading.Channels;
using Microsoft.Extensions.Logging.Abstractions;
using VbNet.LanguageServer.Protocol;
using Xunit;

namespace VbNet.LanguageServer.Tests.Integration;

using LspServer = VbNet.LanguageServer.Core.LanguageServer;

public class CancelRequestIntegrationTests : IAsyncDisposable
{
    private readonly TestTransport _transport = new();
    private readonly LspServer _server;

    public CancelRequestIntegrationTests()
    {
        _server = new LspServer(_transport, NullLoggerFactory.Instance);
        _server.Dispatcher.RegisterRequest<object?, object?>("test/slowRequest", async (_, ct) =>
        {
            await Task.Delay(TimeSpan.FromSeconds(30), ct);
            return new { ok = true };
        });
    }

    [Fact]
    public async Task CancelRequest_CancelsLongRunningRequest()
    {
        using var runCts = new CancellationTokenSource();
        var runTask = _server.RunAsync(runCts.Token);

        _transport.EnqueueMessage("""{"jsonrpc":"2.0","id":1,"method":"test/slowRequest"}""");
        _transport.EnqueueMessage("""{"jsonrpc":"2.0","method":"$/cancelRequest","params":{"id":1}}""");

        var response = await _transport.WaitForSentMessageAsync();
        using var doc = JsonDocument.Parse(response);

        Assert.Equal(1, doc.RootElement.GetProperty("id").GetInt32());
        var error = doc.RootElement.GetProperty("error");
        Assert.Equal(JsonRpcErrorCodes.RequestCancelled, error.GetProperty("code").GetInt32());

        _transport.Complete();
        runCts.Cancel();

        await runTask;
    }

    public async ValueTask DisposeAsync()
    {
        await _server.DisposeAsync();
    }

    private sealed class TestTransport : ITransport
    {
        private readonly Channel<string?> _inbound = Channel.CreateUnbounded<string?>();
        private readonly TaskCompletionSource<string> _sentMessage = new(TaskCreationOptions.RunContinuationsAsynchronously);

        public Task StartAsync(CancellationToken cancellationToken = default) => Task.CompletedTask;

        public async Task<string?> ReadMessageAsync(CancellationToken cancellationToken = default)
        {
            var message = await _inbound.Reader.ReadAsync(cancellationToken);
            return message;
        }

        public Task WriteMessageAsync(string message, CancellationToken cancellationToken = default)
        {
            _sentMessage.TrySetResult(message);
            return Task.CompletedTask;
        }

        public void EnqueueMessage(string message)
        {
            _inbound.Writer.TryWrite(message);
        }

        public Task<string> WaitForSentMessageAsync()
        {
            return _sentMessage.Task;
        }

        public void Complete()
        {
            _inbound.Writer.TryWrite(null);
            _inbound.Writer.TryComplete();
        }

        public ValueTask DisposeAsync() => ValueTask.CompletedTask;
    }
}

File: test/VbNet.LanguageServer.Tests.Vb/Integration/CancelRequestIntegrationTests.vb
Imports System
Imports System.Text.Json
Imports System.Threading
Imports System.Threading.Channels
Imports System.Threading.Tasks
Imports Microsoft.Extensions.Logging.Abstractions
Imports VbNet.LanguageServer.Protocol
Imports Xunit

Namespace VbNet.LanguageServer.Tests.Integration

    Public Class CancelRequestIntegrationTests
        Implements IAsyncDisposable

        Private ReadOnly _transport As New TestTransport()
        Private ReadOnly _server As Core.LanguageServer

        Public Sub New()
            _server = New Core.LanguageServer(_transport, NullLoggerFactory.Instance)
            _server.Dispatcher.RegisterRequest(Of Object, Object)("test/slowRequest", Async Function(unused, ct)
                                                                                         Await Task.Delay(TimeSpan.FromSeconds(30), ct)
                                                                                         Return New With {.ok = True}
                                                                                     End Function)
        End Sub

        <Fact>
        Public Async Function CancelRequest_CancelsLongRunningRequest() As Task
            Using runCts As New CancellationTokenSource()
                Dim runTask = _server.RunAsync(runCts.Token)

                _transport.EnqueueMessage("{""jsonrpc"":""2.0"",""id"":1,""method"":""test/slowRequest""}")
                _transport.EnqueueMessage("{""jsonrpc"":""2.0"",""method"":""$/cancelRequest"",""params"":{""id"":1}}")

                Dim response = Await _transport.WaitForSentMessageAsync()
                Using doc = JsonDocument.Parse(response)
                    Assert.Equal(1, doc.RootElement.GetProperty("id").GetInt32())
                    Dim [error] = doc.RootElement.GetProperty("error")
                    Assert.Equal(JsonRpcErrorCodes.RequestCancelled, [error].GetProperty("code").GetInt32())
                End Using

                _transport.Complete()
                runCts.Cancel()

                Await runTask
            End Using
        End Function

        Public Function DisposeAsync() As ValueTask Implements IAsyncDisposable.DisposeAsync
            Return _server.DisposeAsync()
        End Function

        Private NotInheritable Class TestTransport
            Implements ITransport

            Private ReadOnly _inbound As Channel(Of String) = Channel.CreateUnbounded(Of String)()
            Private ReadOnly _sentMessage As TaskCompletionSource(Of String) =
                New TaskCompletionSource(Of String)(TaskCreationOptions.RunContinuationsAsynchronously)

            Public Function StartAsync(Optional cancellationToken As CancellationToken = Nothing) As Task Implements ITransport.StartAsync
                Return Task.CompletedTask
            End Function

            Public Async Function ReadMessageAsync(Optional cancellationToken As CancellationToken = Nothing) As Task(Of String) Implements ITransport.ReadMessageAsync
                Dim message = Await _inbound.Reader.ReadAsync(cancellationToken)
                Return message
            End Function

            Public Function WriteMessageAsync(message As String, Optional cancellationToken As CancellationToken = Nothing) As Task Implements ITransport.WriteMessageAsync
                _sentMessage.TrySetResult(message)
                Return Task.CompletedTask
            End Function

            Public Sub EnqueueMessage(message As String)
                _inbound.Writer.TryWrite(message)
            End Sub

            Public Function WaitForSentMessageAsync() As Task(Of String)
                Return _sentMessage.Task
            End Function

            Public Sub Complete()
                _inbound.Writer.TryWrite(Nothing)
                _inbound.Writer.TryComplete()
            End Sub

            Public Function DisposeAsync() As ValueTask Implements IAsyncDisposable.DisposeAsync
                Return ValueTask.CompletedTask
            End Function
        End Class
    End Class

End Namespace
